### DISCUSSION TITLE
Bypassing Google Play Purchases in Offline Games

### DISCUSSION THREAD

**Discussion by RevealedSoulEven** [vfsfitvnm/frida-il2cpp-bridge#459] [Show and tell]
You won't believe, it's been less than 24 hours I'm using il2cpp-frida-bridge as you can see my rant posts of how to use and all [here](https://github.com/vfsfitvnm/frida-il2cpp-bridge/discussions/455) and [here](https://github.com/vfsfitvnm/frida-il2cpp-bridge/discussions/456) .

Anyways, this is how I hooked into the transactionState code and retuned PURCHASED object instead of PURCHASING and FAILED objects which were being returned by Play Store.
Here is what was in dump.cs

<pre>
public enum BillingTransactionState
{
	// Fields
	public Int32 value__; // 0x10

        public const BillingTransactionState Purchasing = 0; //  0x0
	public const BillingTransactionState Purchased = 1; // 0x0
	public const BillingTransactionState Failed = 2; // 0x0
	public const BillingTransactionState Restored = 3; // 0x0
	public const BillingTransactionState Deferred = 4; // 0x0
	public const BillingTransactionState Refunded = 5; // 0x0
}
</pre>

So I basically logged the method I found using `Il2Cpp.trace` and got to know that Play Store was returning `PURACHASING` and then `FAILED`

My logic was to simply:
1. Create a duplicate object of `PURCHASED`
2. Intercept the method responsible to get the purchased state.
3. Return the new object we created.

Here is the code.

<pre>
   console.log("Bypass Payments");
    const VoxelBusters = Il2Cpp.domain.assembly("VoxelBusters.EssentialKit").image;
    const BillingServices = VoxelBusters.class("VoxelBusters.EssentialKit.BillingServicesCore.BillingTransactionBase");
    const get_TransactionState = BillingServices.method("get_TransactionState");
    
    const purchased = VoxelBusters.class("VoxelBusters.EssentialKit.BillingTransactionState").field<Il2Cpp.ValueType>("Purchased").value;
    console.log("purr" + (typeof purchased));
    get_TransactionState.implementation = function(this) {
        var x = this.method("get_TransactionState").invoke();
        console.log(x + (typeof x));
        return purchased;
    };
</pre>

I'm new in `frida-il2cpp-bridge` community, but it's been 6 years of game modding. Amazed to see how game hacking developments are being grown and taken to next level by such great devs. Really appreciated.

Enjoy and Keep Learning😊

**Comment by RevealedSoulEven** [comment]
https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/49313374/583d5289-ac78-4fcf-be78-0a8a0351f87e



Working on Temple Run 2
But needed to replace `onTransactionFailure` method with `onTransactionSucceded`

**Reply by takcd** [reply]
I'm pretty new at coding, it only works on offline games? I think is more complicated when we talk about purchase bypass on server games

**Reply by RevealedSoulEven** [reply]
> I'm pretty new at coding, it only works on offline games? I think is more complicated when we talk about purchase bypass on server games

See it matters. Most probably online games are not vulnerable so you can't bypass purchases in them as they validate from servers and play APIs. But still 0.1% luck if it's vulnerable abd uses offline mechanisms, then you can. But assume it not to be possible.

**Reply by takcd** [reply]
Ye I thought about that. I'm solving some coding problems in a Dragon Ball Legends bot, that executes game task sending direct data, and It is already difficult, imagine bypassing the purschase system that most of the times always has a higher security...

**Comment by YesSirBeast** [comment]
Can we also earn play points when we use this method?

**Reply by RevealedSoulEven** [reply]
No. It only touches game files

**Comment by takcd** [comment]
that was huge

**Comment by JayForger** [comment]
can you make a tutorial on how you did these for non-coders like me?

**Reply by RevealedSoulEven** [reply]
For non coders it's very hard to reverse stuffs. You can see aby tutorial and replicate stuff, but when it comes to finding and making your own code, you'll need to learn a little bit of coding and a good experience in reversing.

**Reply by JayForger** [reply]
Ah i see, thank you for replying though
